apisix/plugins/server-info.lua

```diff
@@ -0,0 +1,275 @@
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local require = require
+local core = require("apisix.core")
+
+local ngx_time = ngx.time
+local ngx_timer_at = ngx.timer.at
+local ngx_worker_id = ngx.worker.id
+local type = type
+
+local plugin_name = "server-info"
+local boot_time = os.time()
+local current_timer
+local schema = {
+    type = "object",
+    additionalProperties = false,
+}
+local attr_schema = {
+    type = "object",
+    properties = {
+        report_interval = {
+            type = "integer",
+            description = "server info reporting interval (unit: second)",
+            default = 60,
+            minimum = 60,
+            maximum = 3600,
+        }
+    }
+}
+
+local internal_status = ngx.shared.internal_status
+if not internal_status then
+    error("lua_shared_dict \"internal_status\" not configured")
+end
+
+
+local _M = {
+    version = 0.1,
+    priority = 990,
+    name = plugin_name,
+    schema = schema,
+}
+
+
+local function uninitialized_server_info()
+    return {
+        etcd_version     = "unknown",
+        hostname         = core.utils.gethostname(),
+        id               = core.id.get(),
+        version          = core.version.VERSION,
+        up_time          = ngx_time() - boot_time,
+        last_report_time = -1,
+    }
+end
+
+
+-- server information will be saved into shared memory only if the key
+-- "server_info" not exist if excl is true.
+local function save(data, excl)
+    local handler = excl and internal_status.add or internal_status.set
+
+    local ok, err = handler(internal_status, "server_info", data)
+    if not ok then
+        if excl and err == "exists" then
+            return true
+        end
+
+        return nil, err
+    end
+
+    return true
+end
+
+
+local function encode_and_save(server_info, excl)
+    local data, err = core.json.encode(server_info)
+    if not data then
+        return nil, err
+    end
+
+    return save(data, excl)
+end
+
+
+local function get()
+    local data, err = internal_status:get("server_info")
+    if err ~= nil then
+        return nil, err
+    end
+
+    if not data then
+        return uninitialized_server_info()
+    end
+
+    local server_info, err = core.json.decode(data)
+    if not server_info then
+        return nil, err
+    end
+
+    server_info.up_time = ngx_time() - boot_time
+    return server_info
+end
+
+
+local function get_server_info()
+    local server_info, err = get()
+    if not server_info then
+        core.log.error("failed to get server_info: ", err)
+        return 500, err
+    end
+
+    return 200, core.json.encode(server_info)
+end
+
+
+local function report()
+    local server_info, err = get()
+    if not server_info then
+        core.log.error("failed to get server_info: ", err)
+        return
+    end
+
+    if server_info.etcd_version == "unknown" then
```

Comment:
What if APISIX runs under standalone mode?

Reply:
Since now server-info plugin only reports the basic info to etcd, when APISIX runs under standalone mode, we don't need to launch the timer.